File: apps/vellano/frontend/src/lib/permissions.ts
export const PERMISSION_CATALOG = [
  "users.manage",
  "settings.mutate",
  "catalogue.mutate",
  "po.raise",
  "stock.receive",
  "stock.transfer",
  "stock.adjust",
  "stock.cost.view",
  "till.sell",
  "till.discount",
  "sales.returns",
  "sales.laybys",
  "sales.deliveries",
  "sales.customers",
  "books.mutate",
  "books.journals",
] as const;

export type PermissionKey = (typeof PERMISSION_CATALOG)[number];

export type PermissionHolder = {
  permissions?: readonly string[];
} | null | undefined;

export function hasPermission(
  permissions: readonly string[] | undefined,
  key: string,
): boolean {
  return permissions?.includes(key) === true;
}

export function can(user: PermissionHolder, key: string): boolean {
  return hasPermission(user?.permissions, key);
}

export function canManageLocations(user: PermissionHolder): boolean {
  return can(user, "stock.receive");
}

export function canMutateCatalogue(user: PermissionHolder): boolean {
  return can(user, "catalogue.mutate");
}

export function canRaisePo(user: PermissionHolder): boolean {
  return can(user, "po.raise");
}

export function canReceive(user: PermissionHolder): boolean {
  return can(user, "stock.receive");
}

export function canTransfer(user: PermissionHolder): boolean {
  return can(user, "stock.transfer");
}

export function canReceiveTransfer(user: PermissionHolder): boolean {
  return can(user, "stock.transfer") || can(user, "till.sell");
}

export function canUseTill(user: PermissionHolder): boolean {
  return can(user, "till.sell");
}

export function canMutateBooks(user: PermissionHolder): boolean {
  return can(user, "books.mutate");
}

export function canViewCostAudit(user: PermissionHolder): boolean {
  return can(user, "stock.cost.view");
}

export function canMutateSettings(user: PermissionHolder): boolean {
  return can(user, "settings.mutate");
}

export function canMutateReturns(user: PermissionHolder): boolean {
  return can(user, "sales.returns");
}

export function canMutateLaybys(user: PermissionHolder): boolean {
  return can(user, "sales.laybys");
}

export function canMutateCustomers(user: PermissionHolder): boolean {
  return can(user, "sales.customers");
}

export function canManageCustomerCredit(user: PermissionHolder): boolean {
  return can(user, "users.manage") || can(user, "po.raise");
}

export function canMutateDeliveries(user: PermissionHolder): boolean {
  return can(user, "sales.deliveries");
}

export function canMutatePicks(user: PermissionHolder): boolean {
  return can(user, "stock.transfer") || can(user, "till.sell") || can(user, "sales.deliveries");
}

File: apps/vellano/backend/app/permissions.py
from __future__ import annotations

from typing import Optional

USERS_MANAGE = "users.manage"
SETTINGS_MUTATE = "settings.mutate"
CATALOGUE_MUTATE = "catalogue.mutate"
PO_RAISE = "po.raise"
STOCK_RECEIVE = "stock.receive"
STOCK_TRANSFER = "stock.transfer"
STOCK_ADJUST = "stock.adjust"
STOCK_COST_VIEW = "stock.cost.view"
TILL_SELL = "till.sell"
TILL_DISCOUNT = "till.discount"
SALES_RETURNS = "sales.returns"
SALES_LAYBYS = "sales.laybys"
SALES_DELIVERIES = "sales.deliveries"
SALES_CUSTOMERS = "sales.customers"
BOOKS_MUTATE = "books.mutate"
BOOKS_JOURNALS = "books.journals"
NIA_USE = "nia.use"
NIA_ADMIN = "nia.admin"

PERMISSION_CATALOG: tuple[str, ...] = (
    USERS_MANAGE,
    SETTINGS_MUTATE,
    CATALOGUE_MUTATE,
    PO_RAISE,
    STOCK_RECEIVE,
    STOCK_TRANSFER,
    STOCK_ADJUST,
    STOCK_COST_VIEW,
    TILL_SELL,
    TILL_DISCOUNT,
    SALES_RETURNS,
    SALES_LAYBYS,
    SALES_DELIVERIES,
    SALES_CUSTOMERS,
    BOOKS_MUTATE,
    BOOKS_JOURNALS,
    NIA_USE,
    NIA_ADMIN,
)

PERMISSION_CATALOG_SET: frozenset[str] = frozenset(PERMISSION_CATALOG)

SLUG_OWNER = "owner"
SLUG_BUYER = "buyer"
SLUG_WAREHOUSE = "warehouse"
SLUG_TILL = "till"
SLUG_BOOKS = "books"

ROLE_PRESET_NAMES: dict[str, str] = {
    SLUG_OWNER: "Owner",
    SLUG_BUYER: "Buyer",
    SLUG_WAREHOUSE: "Warehouse",
    SLUG_TILL: "Till",
    SLUG_BOOKS: "Books",
}

ROLE_PRESETS: dict[str, frozenset[str]] = {
    SLUG_OWNER: frozenset(PERMISSION_CATALOG),
    SLUG_BUYER: frozenset({CATALOGUE_MUTATE, PO_RAISE, STOCK_COST_VIEW, NIA_USE}),
    SLUG_WAREHOUSE: frozenset(
        {STOCK_RECEIVE, STOCK_TRANSFER, STOCK_ADJUST, SALES_RETURNS, SALES_DELIVERIES, NIA_USE}
    ),
    SLUG_TILL: frozenset(
        {
            TILL_SELL,
            TILL_DISCOUNT,
            SALES_RETURNS,
            SALES_LAYBYS,
            SALES_DELIVERIES,
            SALES_CUSTOMERS,
            NIA_USE,
        }
    ),
    SLUG_BOOKS: frozenset(
        {BOOKS_MUTATE, BOOKS_JOURNALS, SALES_CUSTOMERS, STOCK_COST_VIEW, NIA_USE}
    ),
}

SYSTEM_ROLE_SLUGS: tuple[str, ...] = (
    SLUG_OWNER,
    SLUG_BUYER,
    SLUG_WAREHOUSE,
    SLUG_TILL,
    SLUG_BOOKS,
)


def role_slug(role: object) -> str:
    value = getattr(role, "value", role)
    return str(value)


def is_known_permission(key: str) -> bool:
    return key in PERMISSION_CATALOG_SET


def validate_permission_keys(keys: list[str]) -> list[str]:
    unknown = [key for key in keys if not is_known_permission(key)]
    if unknown:
        raise ValueError(f"Unknown permission keys: {', '.join(sorted(set(unknown)))}")
    return sorted(set(keys))


def slugify_role_name(name: str, suffix: Optional[int] = None) -> str:
    chars: list[str] = []
    prev_dash = False
    for ch in name.strip().lower():
        if ch.isalnum():
            chars.append(ch)
            prev_dash = False
        elif not prev_dash:
            chars.append("-")
            prev_dash = True
    slug = "".join(chars).strip("-") or "role"
    if suffix is not None:
        slug = f"{slug}-{suffix}"
    return slug[:32]
